sdk/python/feast/offline_server.py

```diff
@@ -127,69 +132,188 @@ def do_get(self, context, ticket):
+            table = self.flights[key]
+            self.offline_store.offline_write_batch(
+                self.store.config, feature_views[0], table, command["progress"]
             )
+        except Exception as e:
+            logger.exception(e)
+            traceback.print_exc()
+            raise e
 
-            try:
-                training_df = (
-                    self.store._get_provider()
-                    .get_historical_features(
-                        config=self.store.config,
-                        feature_views=feature_views,
-                        feature_refs=feature_refs,
-                        entity_df=entity_df,
-                        registry=self.store._registry,
-                        project=project,
-                        full_feature_names=full_feature_names,
-                    )
-                    .to_df()
-                )
-                logger.debug(f"Len of training_df is {len(training_df)}")
-                table = pa.Table.from_pandas(training_df)
-            except Exception as e:
-                logger.exception(e)
-                traceback.print_exc()
-                raise e
+    def write_logged_features(self, command, key):
+        try:
+            table = self.flights[key]
+            feature_service = self.store.get_feature_service(
+                command["feature_service_name"]
+            )
 
-            # Get service is consumed, so we clear the corresponding flight and data
-            del self.flights[key]
+            self.offline_store.write_logged_features(
+                config=self.store.config,
+                data=table,
+                source=FeatureServiceLoggingSource(
+                    feature_service, self.store.config.project
+                ),
+                logging_config=feature_service.logging_config,
+                registry=self.store.registry,
+            )
+        except Exception as e:
+            logger.exception(e)
+            traceback.print_exc()
+            raise e
 
-            return fl.RecordBatchStream(table)
-        else:
-            raise NotImplementedError
+    def pull_all_from_table_or_query(self, command):
+        try:
+            return self._pull_all_from_table_or_query(command).to_df()
+        except Exception as e:
+            logger.exception(e)
+            traceback.print_exc()
+            raise e
+
+    def _pull_all_from_table_or_query(self, command):
+        return self.offline_store.pull_all_from_table_or_query(
+            self.store.config,
+            self.store.get_data_source(command["data_source_name"]),
+            command["join_key_columns"],
+            command["feature_name_columns"],
+            command["timestamp_field"],
+            utils.make_tzaware(datetime.fromisoformat(command["start_date"])),
+            utils.make_tzaware(datetime.fromisoformat(command["end_date"])),
+        )
+
+    def pull_latest_from_table_or_query(self, command):
+        try:
```

Comment:
can we move this `try/catch` block on the `do_get` method and reuse it for all the APIs?

Reply:
Once we do this, we don't need to replicate each API `x` with the equivalent method `_x`